shell/platform/fuchsia/flutter/vsync_waiter.cc

```diff
@@ -134,70 +76,201 @@ fml::TimePoint VsyncWaiter::SnapToNextPhase(
   }
 
-  auto next_vsync_start_time = next_vsync - vsync_offset_;
-
-  if (now >= next_vsync_start_time)
-    next_vsync_start_time =
-        next_vsync_start_time + vsync_info.presentation_interval;
-
-  fml::TimeDelta delta = next_vsync_start_time - now;
-
-  task_runners_.GetUITaskRunner()->PostDelayedTask(
-      [&weak_factory_ui = this->weak_factory_ui_] {
-        if (!weak_factory_ui) {
-          FML_LOG(WARNING) << "WeakPtrFactory for VsyncWaiter is null, likely "
-                              "due to the VsyncWaiter being destroyed.";
-          return;
-        }
-        auto self = weak_factory_ui->GetWeakPtr();
-        if (self) {
-          self->FireCallbackWhenSessionAvailable();
-        }
-      },
-      delta);
+  pair.first = frame_start_time;
+  pair.second = frame_end_time;
+
+  // Useful knowledge for analyzing traces.
+  fml::TimePoint previous_vsync = next_vsync - vsync_interval;
+  TRACE_DURATION(
+      "flutter", "VsyncWaiter::GetTargetTimes", "previous_vsync(ms)",
+      previous_vsync.ToEpochDelta().ToMilliseconds(), "last_targetted(ms)",
+      last_targetted_vsync.ToEpochDelta().ToMilliseconds(), "now(ms)",
+      fml::TimePoint::Now().ToEpochDelta().ToMilliseconds(), "next_vsync(ms))",
+      next_vsync.ToEpochDelta().ToMilliseconds(), "frame_start_time(ms)",
+      frame_start_time.ToEpochDelta().ToMilliseconds(), "frame_end_time(ms)",
+      frame_end_time.ToEpochDelta().ToMilliseconds());
+
+  return pair;
 }
 
-void VsyncWaiter::FireCallbackWhenSessionAvailable() {
-  TRACE_EVENT0("flutter", "VsyncWaiter::FireCallbackWhenSessionAvailable");
-  FML_DCHECK(task_runners_.GetUITaskRunner()->RunsTasksOnCurrentThread());
-  if (session_wait_.Begin(async_get_default_dispatcher()) != ZX_OK) {
-    FML_LOG(ERROR) << "Could not begin wait for Vsync.";
+VsyncWaiter::VsyncWaiter(SessionConnection* session_connection,
+                         flutter::TaskRunners task_runners,
+                         fml::TimeDelta vsync_offset)
+    : flutter::VsyncWaiter(task_runners),
+      session_connection_(session_connection),
+      vsync_offset_(vsync_offset),
+      weak_factory_ui_(nullptr),
+      weak_factory_(this) {
+  FML_CHECK(session_connection_);
+
+  // Generate a WeakPtrFactory for use with the UI thread. This does not need
+  // to wait on a latch because we only ever use the WeakPtrFactory on the UI
+  // thread so we have ordering guarantees (see ::AwaitVSync())
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runners_.GetUITaskRunner(), fml::MakeCopyable([this]() mutable {
+        this->weak_factory_ui_ =
+            std::make_unique<fml::WeakPtrFactory<VsyncWaiter>>(this);
+      }));
+
+  // Initialize the callback that will run every time the SessionConnection
+  // receives one or more frame presented events on Vsync.
+  session_connection_->InitializeVsyncWaiterCallback([this]() {
+    task_runners_.GetUITaskRunner()->PostTask([&weak_factory_ui =
```

Comment:
It is not neccesary to rethread this callback to the UI thread.

This callback bottoms out in VsyncWaiter::FireCallback.  If you read the implementation of it, it protects variables w/ a mutex, and also performs the necesary re-threading internally.

So, 3 things:
1.  Can you update the common VsyncWaiter documentation to indicate that FireCallback may be called from any thread safely?
2. Session callbacks happen on the raster thread.  So change "weak_factory_ui" to "weak_factory_raster" + initialize and destroy it on the raster thread instead of UI.  I plan to move the scenic::Session to the platform thread soon (the thread that weak_factory_ is on), and then the 2nd weak factory will vanish entirely.
3. Since all of this code now runs on the raster thread, and the presentation callbacks + Present also run on the raster thread, VsyncRecorder has no more reason to exist.  Everything happens on one thread.  See my other comment on vsync_recorder.h

Reply:
As we discussed, this is not currently possible due to decreased performance. I filed https://github.com/flutter/flutter/issues/81984 to track this issue.